cmd/libsnap-confine-private/cgroup-support.c

```diff
@@ -97,3 +98,123 @@ bool sc_cgroup_is_v2() {
+            /* we don't want to go up or process the current directory again */
+            continue;
+        }
+        if (sc_streq(ent->d_name, skip)) {
+            continue;
+        }
+        if (sc_startswith(ent->d_name, prefix)) {
+            debug("found matching prefix in \"%s\"", ent->d_name);
+            /* the directory starts with our prefix */
+            return true;
+        }
+        int entfd = openat(dirfd(root), ent->d_name, O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+        if (entfd == -1) {
+            die("cannot open directory entry \"%s\"", ent->d_name);
+        }
+        debug("got directory fd: %d", entfd);
+        /* takes ownership o the file descriptor? */
+        DIR *entdir SC_CLEANUP(sc_cleanup_closedir) = fdopendir(entfd);
+        if (entdir == NULL) {
+            die("cannot fdopendir directory \"%s\"", ent->d_name);
+        }
+        debug("descend into %s", ent->d_name);
+        int found = traverse_looking_for_prefix_in_dir(entdir, prefix, skip);
+        if (found == true) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool sc_cgroup_v2_is_tracking_snap(const char *snap_instance) {
+    debug("is cgroup tracking snap %s?", snap_instance);
+    char tracking_group_name[PATH_MAX] = {0};
+    sc_must_snprintf(tracking_group_name, sizeof tracking_group_name, "snap.%s.", snap_instance);
+
+    char *own_group SC_CLEANUP(sc_cleanup_string) = sc_cgroup_v2_own_path_full();
+    if (own_group == NULL) {
+        die("cannot obtain own group path");
+    }
+    debug("own group: %s", own_group);
+    char *just_leaf = strrchr(own_group, '/');
+    if (just_leaf == NULL) {
+        die("cannot obtain the leaf group path");
+    }
+    /* pointing at /, advance to the next char */
+    just_leaf += 1;
+    debug("leap group: %s", just_leaf);
+
+    // this would otherwise be inherently racy, but the caller is expected to
+    // keep the snap instance lock, thus preventing new apps of that snap from
+    // starting; not we can still return false positive if the currently running
+    // process exits but we look at the hierarchy before systemd has cleaned up
+    // the group
+
+    bool found = false;
+    debug("opening %s", cgroup_dir);
+    DIR *root SC_CLEANUP(sc_cleanup_closedir) = opendir(cgroup_dir);
+    if (root == NULL) {
+        die("cannot open cgroup root dir");
+    }
+    found = traverse_looking_for_prefix_in_dir(root, tracking_group_name, just_leaf);
+    return found;
+}
+
+static const char *self_cgroup = "/proc/self/cgroup";
+
+char *sc_cgroup_v2_own_path_full(void) {
+    FILE *in SC_CLEANUP(sc_cleanup_file) = fopen(self_cgroup, "r");
+    if (in == NULL) {
+        die("cannot open %s", self_cgroup);
+    }
+
+    char *own_group = NULL;
+
+    while (true) {
+        char *line SC_CLEANUP(sc_cleanup_string) = NULL;
+        size_t linesz = 0;
+        ssize_t sz = getline(&line, &linesz, in);
```

Comment:
I find `fgets()` easier to use (if we have a limit for the line length), and it could save you a couple of lines here, but it's my personal preference so no need to change if you disagree :-)